elements/lisk-p2p/test/utils/network_setup.ts

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright © 2019 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+import { P2P } from '../../src/index';
+import { wait } from './helpers';
+import { platform } from 'os';
+export const NETWORK_START_PORT = 5000;
+export const NETWORK_PEER_COUNT = 10;
+export const POPULATOR_INTERVAL = 50;
+export const DEFAULT_MAX_OUTBOUND_CONNECTIONS = 20;
+export const DEFAULT_MAX_INBOUND_CONNECTIONS = 100;
+export const DEFAULT_CONNECTION_TIMEOUT = 500;
+export const DEFAULT_ACK_TIMEOUT = 500;
+export const RATE_CALCULATION_INTERVAL = 10000;
+export const WEB_SOCKET_ENGINE = 'ws';
+export const SEED_PEER_IP = '127.0.0.1';
+export const NETWORK_CREATION_WAIT_TIME = 1000;
+export const NETWORK_DESTROY_WAIT_TIME = 1000;
+
+export const nodeInfoConstants = {
+	nethash: 'da3ed6a45429278bac2666961289ca17ad86595d33b31037615d4b8e8f158bba',
+	version: '1.0.1',
+	protocolVersion: '1.1',
+	minVersion: '1.0.0',
+	os: platform(),
+	height: 0,
+	broadhash: '2768b267ae621a9ed3b3034e2e8a1bed40895c621bbb1bbd613d92b9d24e54b5',
+	nonce: `O2wTkjqplHII`,
+};
+
+interface TestNetworkConfig {
+	networkSize?: number;
+	startNodePort?: number;
+	networkCreationWaitTime?: number;
+	customConfig?: (
+		index: number,
+		startPort: number,
+		networkSize: number,
+	) => object;
+	customNodeInfo?: (
+		index: number,
+		startPort: number,
+		networkSize: number,
+	) => object;
+	customSeedPeers?: (
+		index: number,
+		startPort: number,
+		networkSize: number,
+	) => any[];
+}
+
+export const createNetwork = async ({
+	networkSize,
+	startNodePort,
+	networkCreationWaitTime,
+	customConfig,
+	customNodeInfo,
+	customSeedPeers,
```

Comment:
The reason for injecting these functions is the way we create config for each p2p instance in the network. I saw that we are at most using `index`, `startPort` and `networkSize` to generate some config values. So it's easier to generate any custom config value using these params. I understand that a new developer at first won't be able to understand what are these custom functions. But if any developer wants to add any new integration test, we assume that it has basic knowledge basic working of p2p-lib, about `P2PConfig` and creating a `P2P` instance. 

Also, if a new developer wants to create a new integration test, it will be hard for him to know the consistent `P2Pconfig` to be used because then it has to go through many files to know what config other tests are using (especially with so many hardcoded values and sometimes doesn't make sense and unrelated to a test case). With a single file where all the networks are being created, it's easier for it to understand how it works and it doesn't need to know what other tests are doing. So IMO, even though we are giving up some readability but we are gaining more flexibility, consistency, and scalability. Also, it was actually harder for me to read a test_scenario while debugging when the config code was taking the most chunk of a file.

The other advantage is, we have been dealing with a lot of issues related to unstable integration tests. From my experience, it was hard to find out what is different in the config in a test case that is making it fail. With this approach we are sure, what is the exact difference for a test case.

Reply:
I feel that the `createNetwork` function is currently almost complex enough to require a test case of its own; which is a red flag. Test logic should be very obvious.

Reply:
One problem in Lisk core network test was that often we didn't know if the failing test was because of our lisk-core code or because we messed something up in the test util... Very often there was something wrong in the util.

Reply:
I prefer to not have `customSeedPeers` as that boilerplate is not too much overhead on each test and the value add of the function doesn't justify the added complexity. I prefer @jondubois suggestion of having separate `customNodeInfo`, `customConfig` utility functions. `destroyNetwork` is nice and I think we can keep that as is.

Reply:
So I have simplified to only one function for `customConfig` as per suggested by @jondubois . It's still open for discussion if we think it's more complex to use then we can move the logic to each and every file. But by looking at each test, I think there is not much change in the config and its not so complex to use.